src/Symfony/Component/Scheduler/Generator/MessageGeneratorInterface.php

```diff
@@ -11,10 +11,14 @@
 
 namespace Symfony\Component\Scheduler\Generator;
 
+use Symfony\Component\Scheduler\Schedule;
+
 interface MessageGeneratorInterface
 {
     /**
      * @return iterable<MessageContext, object>
      */
     public function getMessages(): iterable;
+
+    public function getSchedule(): Schedule;
```

Comment:
I'm uncomfortable adding this method to the interface as a MessageGenerator implementation might not use a Schedule.
In any case, I don't see a reason to interact with the schedule before or after handling a message; it doesn't seem right (especially as we now - in 6.4 - have other ways) and again a Schedule is not a requirement to send scheduled messages.
I would recommend removing the Schedule from the events.

Reply:
This is the case when a message is scheduled to be sent every x hours until an event occurs. 
It cannot be processed in the handler (messenger case), because it cannot be already processed at that time. The time at which its state (to be processed or not) can change lies within this x-hour interval. The aim, for example in pre_run, is to be able to check (before sending it to the handler) whether it is still relevant or whether, on the contrary, it should be removed from the stack of messages to be processed (since, for example, it will not have to be sent again if it has been processed in the meantime). The aim is to have access to the schedule and to be able to use the remove method, for example, which will recalculate the messages to be processed (using the setRestart introduced in 6.4). 

Alternatively, I could inject the ScheduleProviderInterface directly into the event ?

Sorry if I misunderstood your point/comment